static/js/table-of-contents.js

```diff
@@ -0,0 +1,35 @@
+function clearActiveStatesInTableOfContents() {
+	document.querySelectorAll('#TableOfContents ul li a').forEach((section) => {
+		section.classList.remove('active');
+	});
+}
+
+window.addEventListener('DOMContentLoaded', () => {
+
+	const observer = new IntersectionObserver(entries => {
+		var currentIR = 0;
+		entries.forEach(entry => {
+			const id = entry.target.getAttribute('id');
+			if (entry.intersectionRatio > currentIR) {
```

Comment:
is it simpler to activate an entry if `entry.intersectionRatio > 0` and deactivate it if `entry.intersecitonRatio === 0`, like [this snippet](https://css-tricks.com/sticky-table-of-contents-with-scrolling-active-states/)? Currently this is doing O(n<sup>2</sup>) (`entries.forEach` then `document.querySelectorAll`) so it's a bit scary 🤣

Reply:
Yeah it works fine for me to just check for `entry.intersectionRatio > 0`
![image](https://user-images.githubusercontent.com/7024160/85604159-9bdacc00-b683-11ea-9458-faac5ec7311b.png)

But for some reason the `Markdown Syntax Guide` generates some weird TOC aligning.

Reply:
@kentnek can you post a code snippet showing how you refactored this logic to use `entry.intersectionRation > 0`?

Reply:
here you go
```js
window.addEventListener('DOMContentLoaded', () => {
  const observer = new IntersectionObserver(entries => {
    entries.forEach(entry => {
      const id = entry.target.getAttribute('id');
      if (entry.intersectionRatio > 0) {
        document.querySelector(`#TableOfContents ul li a[href="#${id}"]`).classList.add('active');
      } else {
        document.querySelector(`#TableOfContents ul li a[href="#${id}"]`).classList.remove('active');
      }
    });
  });

  // Track all headers that have an `id` applied
  document.querySelectorAll('article h3[id], article h2[id]').forEach((section) => {
    observer.observe(section);
  });
});

// Post title doesn't get parsed as it is not part of the content file
(function addPostTitle2ToC() {
  const toc = document.getElementById('TableOfContents');
  if (!toc) return;
  const title = document.getElementById("post__title").innerText;
  const ul = toc.querySelector('ul');
  ul.outerHTML = `<ul><li><a href="#post__title" class="active">${title}</a>${ul.outerHTML}</li></ul>`;
})();
```

also I just realized that your js file was indented with Tab. Please indent with 2 spaces 😛

Reply:
Ok I like this, and it removes the need for the `clearActiveStatesInTableOfContents` function.

> Please indent with 2 spaces

We def need a prettier config up in this project.

Reply:
@jakewies pls help setting that up as I'm also too reliant on my IDE's autoformatting capabilities.

Reply:
Yup I'll make it a priority after this is taken care of.

Reply:
> here you go
> 
> ```js
> window.addEventListener('DOMContentLoaded', () => {
>   const observer = new IntersectionObserver(entries => {
>     entries.forEach(entry => {
>       const id = entry.target.getAttribute('id');
>       if (entry.intersectionRatio > 0) {
>         document.querySelector(`#TableOfContents ul li a[href="#${id}"]`).classList.add('active');
>       } else {
>         document.querySelector(`#TableOfContents ul li a[href="#${id}"]`).classList.remove('active');
>       }
>     });
>   });
> 
>   // Track all headers that have an `id` applied
>   document.querySelectorAll('article h3[id], article h2[id]').forEach((section) => {
>     observer.observe(section);
>   });
> });
> 
> // Post title doesn't get parsed as it is not part of the content file
> (function addPostTitle2ToC() {
>   const toc = document.getElementById('TableOfContents');
>   if (!toc) return;
>   const title = document.getElementById("post__title").innerText;
>   const ul = toc.querySelector('ul');
>   ul.outerHTML = `<ul><li><a href="#post__title" class="active">${title}</a>${ul.outerHTML}</li></ul>`;
> })();
> ```
> 
> @xvallspl also I just realized that your js file was indented with Tab. Please indent with 2 spaces 😛

Like this, it can happen that we are highlighting two sections at the same time instead of only the active one, which is not a problem per se, but we can do better. I started taking a look after your comment yesterday, but I didn't have enough time. Maybe tonight...

Reply:
@xvallspl I'm ok with how it's behaving at the moment, but we can definitely optimize. The optimization doesn't need to block the PR though. We can merge it in it's current state and include a future PR for the optimization.